Formula/python@3.11.rb

```diff
@@ -145,13 +140,19 @@ def install
       --without-ensurepip
       --enable-loadable-sqlite-extensions
       --with-openssl=#{Formula["openssl@1.1"].opt_prefix}
-      --enable-optimizations
       --with-system-expat
       --with-system-ffi
       --with-system-libmpdec
       --with-readline=editline
     ]
 
+    # The --enable-optimization and --with-lto flags diverge from what upstream
+    # python does for their macOS binary releases. They have chosen not to apply
+    # these flags because they want one build that will work across many macOS
+    # releases. Homebrew is not so constrained because the bottling
+    # infrastructure specializes for each macOS major release.
+    args << "--enable-optimizations" if ENV.compiler != :gcc || DevelopmentTools.gcc_version(ENV.cc) >= 5
```

Comment:
New thought:
```suggestion
    args << "--enable-optimizations" if build.bottle?
```
This should effectively fix the problem you're having, and saves us from carrying around the unwieldy compiler check.


My thinking is: extra optimisations are nice for bottles, since saving lots of users little bits of time individually adds up to enough to outweigh the added build time/extra risk of build failure.


For users building a formula for use only on their own machine, the trade-off seems much less worth it. (We also do a similar thing in other formulae: llvm, node, etc.)

Reply:
Seems weird to me to make bottle and from source builds behave differently. The problem is the GCC version, not whether or not a bottle is being built, so while this would technically fix it for me, it doesn't seem like the right approach.

Reply:
> Seems weird to me to make bottle and from source builds behave differently.

This already occurs in multiple places, even inside `brew`. For example, `brew` passes different `-march` flags on Linux depending on whether you are building a bottle or building from source.

Moreover, using a different compiler will already cause the a change in behaviour similar to omitting/including the `--enable-optimizations` flag. If we're okay with that, then we should be ok with using `build.bottle?` here. If we're not okay with using `build.bottle?` here, then I don't see why we should be okay with allowing the use of a compiler different from the one we used to build the bottle.

Reply:
> For example, `brew` passes different `-march` flags on Linux depending on whether you are building a bottle or building from source.

Isn't that for exactly the opposite reason though, to make Linux bottles more portable, i.e. *not* passing `-march` when building a bottle?

> If we're not okay with using `build.bottle?` here, then I don't see why we should be okay with allowing the use of a compiler different from the one we used to build the bottle.

When building from source on Linux, you are most likely not using the same compiler that would have built a Linux bottle anyways. I just don't see the reason for disabling optimizations when building from source, since that's the case where it doesn't need to be portable anyways.

Reply:
> Isn't that for exactly the opposite reason though, to make Linux bottles more portable, i.e. _not_ passing `-march` when building a bottle?

We do pass different `-march` flags for portability, but I'm not sure what you mean by the exactly opposite reason. `--enable-optimizations` here enables PGO, which has no effect on bottle portability.

Note that we _do_ pass `-march=core2` when building a bottle (and on macOS `-march=nehalem`), so we _do_ pass `-march` when building a bottle.

> When building from source on Linux, you are most likely not using the same compiler that would have built a Linux bottle anyways.

Yes, exactly. If we think this:

> Seems weird to me to make bottle and from source builds behave differently.

then we should not allow usage of a different compiler, because different compilers compile code differently.

> I just don't see the reason for disabling optimizations when building from source, since that's the case where it doesn't need to be portable anyways.

As mentioned, disabling optimisations does not affect portability, so not requiring portability when building from source is not relevant. What disabling optimisations does do is:

1. reduce build time
2. reduce the risk of build failures

Both are valuable for users building from source. And I'd argue that both are _more_ valuable than the small benefit of enabling optimisations when building from source.